astropy/io/fits/_tiled_compression/tiled_compression.py

```diff
@@ -260,95 +369,77 @@ def decompress_hdu(hdu):
     tile_shape = _tile_shape(hdu._header)
     data_shape = _data_shape(hdu._header)
 
-    data = np.zeros(data_shape, dtype=BITPIX2DTYPE[hdu._header["ZBITPIX"]])
+    first_array_index = first_tile_index * tile_shape
+    last_array_index = (last_tile_index + 1) * tile_shape
+
+    last_array_index = np.minimum(data_shape, last_array_index)
+
+    buffer_shape = tuple((last_array_index - first_array_index).astype(int))
+
+    data = np.zeros(buffer_shape, dtype=BITPIX2DTYPE[hdu._header["ZBITPIX"]])
 
     quantized = "ZSCALE" in hdu.compressed_data.dtype.names
 
     if len(hdu.compressed_data) == 0:
         return None
 
-    override_itemsize = None
-
-    for irow, tile_slices in enumerate(_iter_array_tiles(data_shape, tile_shape)):
-        row = hdu.compressed_data[irow]
+    if np.product(data.shape) == 0:
+        return data
 
+    for irow, tile_slices in _iter_array_tiles_subset(
+        data_shape, tile_shape, first_tile_index, last_tile_index
+    ):
         # For tiles near the edge, the tile shape from the header might not be
         # correct so we have to pass the shape manually.
         actual_tile_shape = data[tile_slices].shape
-        settings = _header_to_settings(hdu._header, actual_tile_shape)
+        data[tile_slices] = _decompress_tile_common(
+            hdu,
+            row_index=irow,
+            tile_shape=actual_tile_shape,
+            data_shape=data_shape,
+            quantized=quantized,
+        )
 
-        cdata = row["COMPRESSED_DATA"]
+    return data
 
-        # When quantizing floating point data, sometimes the data will not
-        # quantize efficiently. In these cases the raw floating point data can
-        # be losslessly GZIP compressed and stored in the `GZIP_COMPRESSED_DATA`
-        # column.
-        gzip_fallback = len(cdata) == 0
 
-        if gzip_fallback:
-            tile_buffer = _decompress_tile(
-                row["GZIP_COMPRESSED_DATA"], algorithm="GZIP_1"
-            )
+def decompress_hdu(hdu):
```

Comment:
It might be possible to simplify this by just calling the function to extract a section giving it the whole array - need to check if this has any performance impact.

Reply:
Done, no performance impact